File: csharp/tests/CodexWebMcp.Tests/CacheTests.cs
using System;
using System.IO;
using System.Threading.Tasks;
using Xunit;

namespace CodexWebMcp.Tests;

public class CacheTests : IDisposable
{
    private readonly string _tempDir;
    private readonly string? _previousEnv;

    public CacheTests()
    {
        _previousEnv = Environment.GetEnvironmentVariable("CODEX_WEB_CACHE");
        _tempDir = Path.Combine(Path.GetTempPath(), "codex-web-mcp-tests-" + Guid.NewGuid().ToString("N"));
        Directory.CreateDirectory(_tempDir);
        Environment.SetEnvironmentVariable("CODEX_WEB_CACHE", _tempDir);
    }

    public void Dispose()
    {
        Environment.SetEnvironmentVariable("CODEX_WEB_CACHE", _previousEnv);
        try
        {
            if (Directory.Exists(_tempDir)) Directory.Delete(_tempDir, recursive: true);
        }
        catch
        {
            // best-effort cleanup
        }
    }

    [Fact]
    public void KeyHash_IsDeterministic()
    {
        var a = Cache.KeyHash("web_search", "hello world");
        var b = Cache.KeyHash("web_search", "hello world");
        Assert.Equal(a, b);
        Assert.Equal(32, a.Length);
    }

    [Fact]
    public void KeyHash_DiffersByTool()
    {
        var a = Cache.KeyHash("a", "x");
        var b = Cache.KeyHash("b", "x");
        Assert.NotEqual(a, b);
    }

    [Fact]
    public void KeyHash_DiffersByPayload()
    {
        var a = Cache.KeyHash("web_search", "alpha");
        var b = Cache.KeyHash("web_search", "beta");
        Assert.NotEqual(a, b);
    }

    [Fact]
    public async Task RoundTrip_Hit()
    {
        const string tool = "web_search";
        const string payload = "round-trip-query";
        const string content = "<untrusted source='codex:web_search'>\nbody\n</untrusted>";

        await Cache.PutAsync(tool, payload, content);
        var got = await Cache.GetAsync(tool, payload, ttlSeconds: 3600);

        Assert.Equal(content, got);
    }

    [Fact]
    public async Task Get_Miss_ReturnsNull()
    {
        var got = await Cache.GetAsync("web_search", "never-written-" + Guid.NewGuid(), ttlSeconds: 3600);
        Assert.Null(got);
    }

    [Fact]
    public async Task Get_Expired_ReturnsNull()
    {
        const string tool = "web_search";
        const string payload = "expired-query";
        const string content = "expired body";

        await Cache.PutAsync(tool, payload, content);

        // Force the file's mtime far into the past.
        var hash = Cache.KeyHash(tool, payload);
        var path = Path.Combine(Config.CacheSubdir, $"{tool}-{hash}.json");
        Assert.True(File.Exists(path));
        File.SetLastWriteTimeUtc(path, DateTime.UtcNow.AddHours(-2));

        var got = await Cache.GetAsync(tool, payload, ttlSeconds: 60);
        Assert.Null(got);
    }

    [Fact]
    public async Task Get_WithinTTL_ReturnsContent()
    {
        const string tool = "web_search";
        const string payload = "fresh-query";
        const string content = "fresh body";

        await Cache.PutAsync(tool, payload, content);

        // Set mtime to 30 seconds ago — well within a 3600s TTL.
        var hash = Cache.KeyHash(tool, payload);
        var path = Path.Combine(Config.CacheSubdir, $"{tool}-{hash}.json");
        File.SetLastWriteTimeUtc(path, DateTime.UtcNow.AddSeconds(-30));

        var got = await Cache.GetAsync(tool, payload, ttlSeconds: 3600);
        Assert.Equal(content, got);
    }
}

File: fsharp/tests/CodexWebMcp.FSharp.Tests/CacheTests.fs
module CodexWebMcp.FSharp.Tests.CacheTests

open System
open System.IO
open Xunit
open CodexWebMcp.FSharp

type CacheFixture() =
    let tempDir =
        Path.Combine(Path.GetTempPath(), "codex-web-mcp-fs-tests-" + Guid.NewGuid().ToString("N"))
    let previous = Environment.GetEnvironmentVariable("CODEX_WEB_CACHE")
    do
        Directory.CreateDirectory(tempDir) |> ignore
        Environment.SetEnvironmentVariable("CODEX_WEB_CACHE", tempDir)

    member _.Dir = tempDir

    interface IDisposable with
        member _.Dispose() =
            Environment.SetEnvironmentVariable("CODEX_WEB_CACHE", previous)
            try Directory.Delete(tempDir, true) with _ -> ()

[<Fact>]
let ``KeyHash is deterministic and 32 hex chars`` () =
    let a = Cache.keyHash "web_search" "hello world"
    let b = Cache.keyHash "web_search" "hello world"
    Assert.Equal(a, b)
    Assert.Equal(32, a.Length)

[<Fact>]
let ``KeyHash differs by tool`` () =
    let a = Cache.keyHash "web_search" "x"
    let b = Cache.keyHash "web_search_raw" "x"
    Assert.NotEqual<string>(a, b)

[<Fact>]
let ``KeyHash differs by payload`` () =
    let a = Cache.keyHash "web_search" "x"
    let b = Cache.keyHash "web_search" "y"
    Assert.NotEqual<string>(a, b)

[<Fact>]
let ``Round trip`` () =
    use fixture = new CacheFixture()
    Cache.put "web_search" "query1" "the answer"
    match Cache.get "web_search" "query1" 3600 with
    | Some v -> Assert.Equal("the answer", v)
    | None -> Assert.Fail("expected Some")

[<Fact>]
let ``Get miss returns None`` () =
    use fixture = new CacheFixture()
    Assert.Equal(None, Cache.get "web_search" "never-written" 3600)

[<Fact>]
let ``Get expired returns None`` () =
    use fixture = new CacheFixture()
    Cache.put "web_search" "stale" "old"
    let file =
        Path.Combine(fixture.Dir, "cache", sprintf "web_search-%s.json" (Cache.keyHash "web_search" "stale"))
    File.SetLastWriteTimeUtc(file, DateTime.UtcNow.AddHours(-2.0))
    Assert.Equal(None, Cache.get "web_search" "stale" 60)

[<Fact>]
let ``Get within TTL returns content`` () =
    use fixture = new CacheFixture()
    Cache.put "web_search" "fresh" "content"
    match Cache.get "web_search" "fresh" 3600 with
    | Some v -> Assert.Equal("content", v)
    | None -> Assert.Fail("expected Some")
